### PR TITLE
Support uneven sharding for FSDP2 + TP

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150393
* #150146
* __->__ #148894

### CODE REVIEW

**Comment by lw** [comment]
CC @fmassa

File: test/distributed/tensor/test_dtensor.py [hunk 1]
```diff
@@ -956,7 +956,7 @@ def test_split_tensor_1D(self) -> None:
             )
             if size == 0:
                 # when tensor size is 0, there is no padding needed for all the ranks.
-                expected_pad_sizes = []
+                expected_pad_sizes = [0] * self.world_size
```
Comment (lw) [inline test/distributed/tensor/test_dtensor.py:959]: I'm changing this calling convention, as it makes for a simpler implementation of `_split_tensor` _and_ for a simpler callsite logic (no more checking whether pad_sizes is empty). Since `_split_tensor` is private I don't believe this is really a BC-change.

File: torch/distributed/tensor/placement_types.py [hunk 2]
```diff
@@ -89,39 +89,22 @@ def _split_tensor(
 
         # chunk tensor over dimension `dim` into n slices
         tensor_list = list(torch.chunk(tensor, num_chunks, dim=self.dim))
-        num_empty_tensors = num_chunks - len(tensor_list)
```
Comment (lw) [inline torch/distributed/tensor/placement_types.py:92]: I'm simplifying the implementation of this `Shard._split_tensor` method, without changing the behavior (except for when we return an empty list for pad_sizes), and then I'll reuse the same structute for `_StridedShard._split_tensor` below.

File: torch/distributed/tensor/placement_types.py [hunk 3]
```diff
@@ -192,8 +175,10 @@ def _shard_tensor(
         )
 
         # Only unpad if the local_tensor was padded on the dimension.
-        if pad_sizes and pad_sizes[mesh_dim_local_rank] > 0:
```
Comment (lw) [inline torch/distributed/tensor/placement_types.py:195]: We don't have to check for non-emptyness anymore thanks to the change in calling conventions.

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

But I'd prefer if someone from the distributed team double-review this PR.

cc @tianyu-l

**Review by weifengpy — APPROVED** [review]
the fsdp2 part looks good. One nit is to verify the correctness of state dict for unevan sharding: https://github.com/pytorch/pytorch/blob/a5f6b24d87bbf735376d5c6d4c8bb2a7abaf0c88/test/distributed/_composable/fsdp/test_fully_shard_state_dict.py#L167

will leave DTensor part to @XilunWu and Tianyu for review. DTensor changes are the hard part

File: torch/distributed/tensor/placement_types.py [hunk 4]
```diff
@@ -465,37 +446,35 @@ def _split_tensor(
...
-                    for j in range(group_size)
-                ],
+        # compute the chunk size inline with ``torch.chunk`` to calculate padding
+        full_chunk_size = (tensor.size(self.dim) + total_split - 1) // total_split
+
+        shard_list: list[torch.Tensor] = []
+        pad_sizes: list[int] = []
+        for i in range(num_chunks):
+            shard = torch.cat(
+                [tensor_list[i + j * num_chunks] for j in range(self.split_factor)],
```
Comment (tianyu-l) [inline torch/distributed/tensor/placement_types.py:467]: I feel the comment `# stride is num_chunks` still helps. Can we keep it somewhere and maybe even make it more noticeable? After all the name of this placement is "strided" sharding.

**Review by tianyu-l — APPROVED** [review]
It looks awesome to me!

**Comment by lw** [comment]
After further local testing I found some issues, will need to keep iterating

**Review by XilunWu — CHANGES_REQUESTED** [review]
I feel this uneven sharding is not compatible with _StridedShard(). _StridedShard is designed to represent a right-to-left sharding pattern for example, `[_StridedShard(dim=0), Shard(0)]` is equivalent to `[Shard(0), Shard(0)]` but sharding tensor dimension 0 over mesh dim-1 first then mesh dim-0. 

A concrete example is sharding tensor `[0, 1, 2, 3, 4]` over a 2-d mesh `[[rank 0, rank 1], [rank 2, rank 3]]`.

A regular `[Shard(0), Shard(0)]` sharding produces: ('#' stands for padding)
rank 0: `[0, 1]`    |     rank 1: `[2, #]`
rank 2: `[3, #]`   |      rank 3: `[4, #]`

A right-to-left `[Shard(0), Shard(0)]` sharding produces:
rank 0: `[0, 1]`    |     rank 1: `[3, #]`
rank 2: `[2, #]`   |      rank 3: `[4, #]`
This is what `[_StridedShard(0, split_factor=2), Shard(0)]` expects.

But this PR's implementation on `[_StridedShard(0, split_factor=2), Shard(0)]` produces:
rank 0: `[0, 1]`    |     rank 1: `[4, #]`
rank 2: `[2, #]`   |      rank 3: `[3, #]`

File: torch/distributed/tensor/placement_types.py [hunk 5]
```diff
@@ -465,37 +446,35 @@ def _split_tensor(
...
-        group_size = self.split_factor
-        total_split_tensor_list = list(torch.chunk(tensor, total_split, dim=self.dim))
-        tensor_list = [
-            torch.cat(
-                [
-                    total_split_tensor_list[i + j * num_chunks]  # stride is num_chunks
-                    for j in range(group_size)
-                ],
+        # compute the chunk size inline with ``torch.chunk`` to calculate padding
+        full_chunk_size = (tensor.size(self.dim) + total_split - 1) // total_split
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:461]: just wondering, can we directly call self._local_shard_size_on_dim for this?

File: torch/distributed/tensor/placement_types.py [hunk 6]
```diff
@@ -465,37 +446,35 @@ def _split_tensor(
         with_padding: bool = True,
         contiguous: bool = True,
     ) -> tuple[list[torch.Tensor], list[int]]:
-        """
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:468]: would be nice to document the intuition behind StridedShard._split_tensor.  IIUC it is 
 - we 'split' in 'Left to Right' order, meaning for [StridedShard(0, sf=2), Shard(0)] we first split the tensor over the strided placment, and then later we split it over the shard placement.
example:

orig tensor:
[0, 1, 2, 3, 4]

Step 1: Strided shard (including strided split and cat)
[0, 1] [2, 3] [4, ] []
[0, 1, 4],  [2, 3]

Step 2: Regular Shard(0) split
Rank0: [0, 1]   Rank1 [2,]
Rank2: [4,]     Rank3 [3]

**Comment by wconstab** [comment]
Closing as replaced by https://github.com/pytorch/pytorch/pull/150490
